sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/JacksonAdapter.java

```diff
@@ -234,96 +263,178 @@ public String serializeList(List<?> list, CollectionFormat format) {
+            this.values = new HashMap<>();
+            this.declaringField = declaringField;
+        }
+
+        boolean headerStartsWithPrefix(String headerName) {
+            return headerName.startsWith(prefix);
+        }
+
+        void addHeader(String headerName, String headerValue) {
+            values.put(headerName.substring(prefixLength), headerValue);
+        }
+
+        void injectValuesIntoDeclaringField(Object deserializedHeaders, ClientLogger logger) {
+            /*
+             * First check if the deserialized headers type has a public setter.
+             */
+            if (usePublicSetter(deserializedHeaders, logger)) {
+                return;
+            }
+
+            logger.verbose("Failed to find or use public setter to set header collection.");
+
+            /*
+             * Otherwise fallback to setting the field directly.
+             */
+            final boolean declaredFieldAccessibleBackup = declaringField.isAccessible();
+            try {
+                if (!declaredFieldAccessibleBackup) {
+                    AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+                        declaringField.setAccessible(true);
+                        return null;
+                    });
+                }
+                declaringField.set(deserializedHeaders, values);
+                logger.verbose("Set header collection by accessing the field directly.");
+            } catch (IllegalAccessException ex) {
+                logger.warning("Failed to inject header collection values into deserialized headers.", ex);
+            } finally {
+                if (!declaredFieldAccessibleBackup) {
+                    AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+                        declaringField.setAccessible(false);
+                        return null;
+                    });
+                }
+            }
+        }
+
+        private boolean usePublicSetter(Object deserializedHeaders, ClientLogger logger) {
+            try {
+                String potentialSetterName = getPotentialSetterName();
+
+                // This could be cached.
+                Method setterMethod = deserializedHeaders.getClass().getDeclaredMethod(potentialSetterName, Map.class);
+                if (Modifier.isPublic(setterMethod.getModifiers())) {
+                    setterMethod.invoke(deserializedHeaders, values);
+                    logger.verbose("User setter %s on class %s to set header collection.", potentialSetterName,
+                        deserializedHeaders.getClass().getSimpleName());
+                    return true;
+                }
+
+                return false;
+            } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException ignored) {
+                return false;
+            }
+        }
+
+        private String getPotentialSetterName() {
+            String fieldName = declaringField.getName();
```

Comment:
Is this guaranteed to be non-null and non-empty?

Reply:
Should be, this is a field retrieved using the `getDeclaredFields()` method on `Class`, so I would be surprised if a null or empty name was returned